### PR TITLE
Improve reduction of terms aggregations

### PR DESCRIPTION
**Pull request by jimczi**
Today, the terms aggregation reduces multiple aggregations at once using a map
to group same buckets together. This operation can be costly since it requires
to lookup every bucket in a global map with no particular order.
This commit changes how term buckets are sorted by shards and partial reduces in
order to be able to reduce results using a merge-sort strategy.
For bwc, results are merged with the legacy code if any of the aggregations use
a different sort (if the shard response was created by a node in prior versions).

Relates #51857

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java [hunk 1]
```diff
@@ -724,8 +727,15 @@ void buildSubAggs(StringTerms.Bucket[][] topBucketsPreOrd) throws IOException {
 
         @Override
         StringTerms buildResult(long owningBucketOrd, long otherDocCount, StringTerms.Bucket[] topBuckets) {
-            return new StringTerms(name, order, bucketCountThresholds.getRequiredSize(), bucketCountThresholds.getMinDocCount(),
-                metadata(), format, bucketCountThresholds.getShardSize(), showTermDocCountError,
+            final BucketOrder reduceOrder;
+            if (isKeyOrder(order) == false) {
+                reduceOrder = isKeyOrder(order) ? InternalOrder.unwrap(order) : InternalOrder.key(true);
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java]: thanks, that's a leftover, I pushed https://github.com/elastic/elasticsearch/pull/61779/commits/99bc538cad46e88ab35a84ecb32db184fd2890e0

File: server/src/main/java/org/elasticsearch/search/aggregations/InternalOrder.java [hunk 2]
```diff
@@ -369,6 +369,20 @@ public static boolean isKeyDesc(BucketOrder order) {
         return isOrder(order, KEY_DESC);
     }
 
+    /**
+     * Return the primary {@link BucketOrder} if the provided <code>order</code>
+     * is a {@link CompoundOrder}.
+     */
+    public static BucketOrder unwrap(BucketOrder order) {
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/InternalOrder.java]: I removed the usage of this method in https://github.com/elastic/elasticsearch/pull/61779/commits/99bc538cad46e88ab35a84ecb32db184fd2890e0

File: server/src/test/java/org/elasticsearch/search/aggregations/InternalAggregationsTests.java [hunk 3]
```diff
@@ -35,6 +35,7 @@
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.test.InternalAggregationTestCase;
 import org.elasticsearch.test.VersionUtils;
+import org.junit.Ignore;
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/search/aggregations/InternalAggregationsTests.java:38]: Leftover?

**Review by nik9000 — APPROVED** [review]
LGTM

**Comment by jimczi** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java [hunk 4]
```diff
@@ -189,44 +220,173 @@ protected final void doWriteTo(StreamOutput out) throws IOException {
     @Override
     public abstract B getBucketByKey(String term);
 
-    @Override
+    private static class IteratorAndCurrent<B extends InternalTerms.Bucket<B>> {
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java]: +1, I pushed https://github.com/elastic/elasticsearch/pull/61779/commits/d752ac7c09cea32aa753be06eb99ed628421d61f

**Comment by jimczi** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

**Comment by jimczi** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

**Pull request merged**